core/src/main/scala/kafka/server/BrokerMetadataEventManager.scala

```diff
@@ -0,0 +1,198 @@
+import org.apache.kafka.common.utils.Time
+import org.apache.kafka.raft.RaftMessage
+
+object BrokerMetadataEventManager {
+  val BrokerMetadataEventThreadName = "broker-metadata-event-thread"
+  val EventQueueTimeMetricName = "EventQueueTimeMs"
+  val EventQueueSizeMetricName = "EventQueueSize"
+}
+
+sealed trait Event {}
+
+final case object StartupEvent extends Event
+final case class MetadataEvent(metadataMessage: RaftMessage) extends Event
+/*
+ * ShutdownEvent is necessary for the case when the event thread is blocked in queue.take() -- this event wakes it up.
+ * Otherwise, this event has no other semantic meaning, it is not guaranteed to be seen by the processing thread,
+ * and it cannot be relied upon to initiate any other functionality; it exists solely for the wakeup case mentioned.
+ */
+final case object ShutdownEvent extends Event
+
+class QueuedEvent(val event: Event, val enqueueTimeMs: Long) {
+  override def toString: String = {
+    s"QueuedEvent(event=$event, enqueueTimeMs=$enqueueTimeMs)"
+  }
+}
+
+trait MetadataEventProcessor {
+  def processStartup(): Unit
+  def process(metadataMessage: RaftMessage): Unit
+  // note the absence of processShutdown(); see ShutdownEvent above for why it does not exist.
+}
+
+class BrokerMetadataEventManager(replicaManager: ReplicaManager,
+                                 groupCoordinator: GroupCoordinator,
+                                 metadataCache: MetadataCache,
+                                 quotas: QuotaManagers,
+                                 clusterId: String,
+                                 time: Time,
+                                 overrideProcessor: Option[MetadataEventProcessor] = None,
+                                 eventQueueTimeTimeoutMs: Long = 300000) extends KafkaMetricsGroup with MetadataEventProcessor {
+
+  import BrokerMetadataEventManager._
+
+  val queue = new LinkedBlockingQueue[QueuedEvent]
+
+  val thread = new BrokerMetadataEventThread(BrokerMetadataEventThreadName)
+
+  val processor: MetadataEventProcessor = overrideProcessor.getOrElse(this)
+
+  // metrics
+  private val eventQueueTimeHist = newHistogram(EventQueueTimeMetricName)
+  newGauge(EventQueueSizeMetricName, () => queue.size)
+
+  def start(): Unit = {
+    put(StartupEvent)
+    thread.start()
+  }
+
+  def close(): Unit = {
+    try {
+      thread.initiateShutdown()
+      put(ShutdownEvent) // wake up the thread in case it is blocked on queue.take()
+      thread.awaitShutdown()
+    } finally {
+      removeMetric(EventQueueTimeMetricName)
+      removeMetric(EventQueueSizeMetricName)
+    }
+  }
+
+  def put(event: Event): QueuedEvent = {
+    val queuedEvent = new QueuedEvent(event, time.milliseconds())
+    queue.put(queuedEvent)
+    queuedEvent
+  }
+
+  private def pollFromEventQueue(): QueuedEvent = {
+    val hasRecordedValue = eventQueueTimeHist.count() > 0
+    if (hasRecordedValue) {
+      val event = queue.poll(eventQueueTimeTimeoutMs, TimeUnit.MILLISECONDS)
```

Comment:
What's the purpose of having a timeout here?  It will use extra cpu, right

Reply:
I got this from ControllerEventManager, which was recently amended to resolve https://issues.apache.org/jira/browse/KAFKA-10189.  The logic resets the metric to 0 if it is non-zero and there are no events within the timeout period.